--- a/options.h
+++ b/options.h
@@ -55,6 +55,7 @@
 	UPNPPCPMINLIFETIME,		/* minimum lifetime for PCP mapping */
 	UPNPPCPMAXLIFETIME,		/* maximum lifetime for PCP mapping */
 	UPNPPCPALLOWTHIRDPARTY,		/* allow third-party requests */
+	UPNPUPNPMAXLIFETIME,		/* maximum lifetime for UPnP IGD mapping */
 #ifdef USE_NETFILTER
 	UPNPTABLENAME,
 	UPNPNATTABLENAME,
--- a/options.c
+++ b/options.c
@@ -85,6 +85,7 @@
 	{ UPNPPCPMINLIFETIME, "min_lifetime"},
 	{ UPNPPCPMAXLIFETIME, "max_lifetime"},
 	{ UPNPPCPALLOWTHIRDPARTY, "pcp_allow_thirdparty"},
+	{ UPNPUPNPMAXLIFETIME, "upnp_max_lifetime"},
 #endif
 	{ UPNPENABLE, "enable_upnp"},
 #ifdef USE_PF
--- a/upnpglobalvars.h
+++ b/upnpglobalvars.h
@@ -50,6 +50,7 @@
 
 extern unsigned long int min_lifetime;
 extern unsigned long int max_lifetime;
+extern unsigned long int upnp_max_lifetime;
 
 /* runtime boolean flags */
 extern int runtime_flags;
--- a/upnpglobalvars.c
+++ b/upnpglobalvars.c
@@ -55,6 +55,7 @@
 #ifdef ENABLE_PCP
 unsigned long int min_lifetime = 120;
 unsigned long int max_lifetime = 86400;
+unsigned long int upnp_max_lifetime = 604800;
 #endif
 
--- a/miniupnpd.c
+++ b/miniupnpd.c
@@ -1519,6 +1519,12 @@
 			case UPNPPCPMAXLIFETIME:
 					max_lifetime = atoi(ary_options[i].value);
 					/* maximum is 24 hours */
 					if (max_lifetime > 86400 ) {
 						max_lifetime = 86400;
 					}
 				break;
+			case UPNPUPNPMAXLIFETIME:
+				upnp_max_lifetime = atoi(ary_options[i].value);
+				if(upnp_max_lifetime < 120) upnp_max_lifetime = 120;
+				if(upnp_max_lifetime > 604800) upnp_max_lifetime = 604800;
+				break;
 			case UPNPPCPALLOWTHIRDPARTY:
 				if(strcmp(ary_options[i].value, "yes") == 0)
 					SETFLAG(PCP_ALLOWTHIRDPARTYMASK);
--- a/upnpsoap.c
+++ b/upnpsoap.c
@@ -548,8 +548,8 @@
 	 * mechanisms like WWW-administration, remote management or local
 	 * management.
 	 * If a control point uses the value 0 to indicate an infinite lease
-	 * time mapping, it is REQUIRED that gateway uses the maximum value
-	 * instead (e.g. 604800 seconds) */
-	if(leaseduration == 0 || leaseduration > 604800)
-		leaseduration = 604800;
+	 * time mapping, use configured upnp_max_lifetime value */
+	if(leaseduration == 0 || leaseduration > upnp_max_lifetime)
+		leaseduration = upnp_max_lifetime;
 #endif
 
